### PR TITLE
Add simple GetAll request to test client/node

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**

### CODE REVIEW

**Comment by gavin-norman-sociomantic** [comment]
This GetAll request was implemented in order to serve as a foundation for experiments with https://github.com/sociomantic-tsunami/swarm/issues/3, but is valuable in its own right so has been submitted as a PR.

**Comment by gavin-norman-sociomantic** [comment]
Try again. (Added missing file ;)

**Comment by gavin-norman-sociomantic** [comment]
Updated with D2 const fix.

File: test/neo/client/Client.d [hunk 1]
```diff
@@ -222,6 +248,95 @@ public class Client
...
             return succeeded;
         }
+
+        /***********************************************************************
+
+            GetAll iterator "fruct" (foreach struct).
+
+        ***********************************************************************/
+
+        private struct GetAllFruct
```
Comment (nemanja-boric-sociomantic) [inline test/neo/client/Client.d:258]: <3

Comment (gavin-norman-sociomantic) [reply test/neo/client/Client.d:258]: Haha. `tango` is dead; long live the `fruct`.

**Comment by nemanja-boric-sociomantic** [comment]
I just did the high-look overview, and it LGTM. I will later do a more careful pass.

**Comment by gavin-norman-sociomantic** [comment]
Rebased.

File: test/neo/client/request/GetAll.d [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    Notification smart union.
+
+*******************************************************************************/
+
+public alias SmartUnion!(NotificationUnion) Notification;
+
+/*******************************************************************************
+
+    Type of notifcation delegate.
```
Comment (nemanja-boric-sociomantic) [inline test/neo/client/request/GetAll.d]: `notification`

File: test/neo/main.d [hunk 3]
```diff
@@ -113,6 +114,50 @@ class Test : Task
...
+        // Add some records.
+        cstring[hash_t] records =
+            [0x1 : "you"[], 0x2 : "say", 23 : "hello"];
+        foreach ( key, value; records )
+        {
+            auto ok = this.client.blocking.put(key, value,
+                ( Client.Neo.Put.Notification info, Client.Neo.Put.Args args )
+                {
+                    formatNotification(info, msg_buf);
+                    Stdout.formatln("Put: {}", msg_buf);
```
Comment (nemanja-boric-sociomantic) [inline test/neo/main.d]: It is the question if the printing out in tests is worth it. I feel `enforce` is enough, since this is just a noise.

**Review by nemanja-boric-sociomantic — APPROVED** [review]
Two small nitpicks, but generally LGTM!

File: test/neo/main.d (hunk 3 above)
Comment (gavin-norman-sociomantic) [reply test/neo/main.d]: Ah yeah, right, I meant to remove them.

**Comment by gavin-norman-sociomantic** [comment]
Updated.

**Comment by gavin-norman-sociomantic** [comment]
Nemanja has reviewed this but has now selfishly left on holiday for 3 weeks. @david-eckardt-sociomantic could you continue the review / merge?

File: test/neo/client/Client.d [hunk 4]
```diff
@@ -222,6 +248,95 @@ public class Client
...
+
+                    with ( info.Active ) switch ( info.active )
+                    {
+                        case record:
+                            res = dg(info.record.key, info.record.value);
+
+                            // This simple iterator implementation does not
+                            // support ending the request via breaking the
+                            // foreach loop.
+                            assert(!res);
```
Comment (david-eckardt-sociomantic) [inline test/neo/client/Client.d:299]: Hahaha, so an assertion will fail if you `break` the `foreach` loop. Given that the `res` variable isn't necessary.

Comment (gavin-norman-sociomantic) [reply test/neo/client/Client.d:299]: Surely that's the only way to tell that the user broke the loop, though?

Comment (david-eckardt-sociomantic) [reply test/neo/client/Client.d:299]: Yes, it is. It just looks a bit funny to me: `break;` seriously breaks something.

File: test/neo/client/request/internal/GetAll.d [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    ***************************************************************************/
+
+    mixin RequestCore!(RequestType.AllNodes, RequestCode.GetAll, 0, Args,
+        SharedWorking, Working, Notification);
+
+    /***************************************************************************
+
+        Request handler. Called from RequestOnConn.runHandler().
+
+        Params:
```
Comment (david-eckardt-sociomantic) [inline test/neo/client/request/internal/GetAll.d:91]: `conn` not documented

File: test/neo/client/Client.d (hunk 4 above)
Comment (gavin-norman-sociomantic) [reply test/neo/client/Client.d:299]: Yeah. It's just an example and it's clearly documented that `break` is not supported.

File: test/neo/client/request/internal/GetAll.d (hunk 5 above)
Comment (gavin-norman-sociomantic) [reply test/neo/client/request/internal/GetAll.d:91]: Updated.

**Comment by gavin-norman-sociomantic** [comment]
Rebased.

**Review by david-eckardt-sociomantic — APPROVED** [review]

**Comment by nemanja-boric-sociomantic** [comment]
> Nemanja has reviewed this but has now selfishly left on holiday for 3 weeks. @david-eckardt-sociomantic could you continue the review / merge?

:angel:

**Pull request merged**